Plutarch/Internal/Lift.hs

```diff
@@ -214,51 +285,83 @@ instance DerivePlutusType (DeriveDataPLiftable a h) where
 -- | @since WIP
 instance
   ( PlutusType a
-  , PInner a ~ PData
+  , PSubtype PData a
   , PTx.ToData h
   , PTx.FromData h
   ) =>
   PLiftable (DeriveDataPLiftable a h)
   where
   type AsHaskell (DeriveDataPLiftable a h) = h
+  type PlutusRepr (DeriveDataPLiftable a h) = PTx.Data
+
+  {-# INLINEABLE toPlutarchRepr #-}
+  toPlutarchRepr = PTx.toData
+
   {-# INLINEABLE toPlutarch #-}
-  toPlutarch = unsafeFromUni @_ @PLCData.Data . PTx.toData
+  toPlutarch = toPlutarchUni
+
+  {-# INLINEABLE fromPlutarchRepr #-}
+  fromPlutarchRepr = PTx.fromData
+
   {-# INLINEABLE fromPlutarch #-}
-  fromPlutarch t = do
-    res <- unsafeToUni t
-    case PTx.fromData res of
-      Nothing -> Left CouldNotDecodeData
-      Just res' -> pure res'
+  fromPlutarch = fromPlutarchUni
 
-{- | Helper for writing 'PLifted' instances. This is /highly/ unsafe, and only
-suitable for internal use!
+{- | @via@-deriving helper, indicating that @wrapper@ has a Haskell-level equivalent
+@h@ by way of encoding of @inner@. It requires that @AsHaskell inner@ has the same
+Haskell representation as @h@
 
 @since WIP
 -}
-unsafeFromUni ::
-  forall (a :: S -> Type) (h :: Type) (s :: S).
-  PLC.DefaultUni `Includes` h =>
-  h ->
-  PLifted a s
-unsafeFromUni =
-  PLifted . punsafeConstantInternal . PLC.someValue @h @PLC.DefaultUni
+newtype DeriveNewtypePLiftable (wrapper :: S -> Type) (inner :: S -> Type) (h :: Type) (s :: S)
+  = DeriveNewtypePLiftable (wrapper s)
+  deriving stock (Generic)
+  deriving anyclass (PlutusType)
+
+-- | @since WIP
+instance DerivePlutusType (DeriveNewtypePLiftable w i h) where
+  type DPTStrat _ = PlutusTypeNewtype
+
+-- | @since WIP
+instance (PLiftable inner, Coercible (AsHaskell inner) h) => PLiftable (DeriveNewtypePLiftable wrapper inner h) where
+  type AsHaskell (DeriveNewtypePLiftable wrapper inner h) = h
+  type PlutusRepr (DeriveNewtypePLiftable wrapper inner h) = PlutusRepr inner
+
+  {-# INLINEABLE toPlutarchRepr #-}
+  toPlutarchRepr = toPlutarchRepr @inner . coerce @h @(AsHaskell inner)
+
+  {-# INLINEABLE toPlutarch #-}
+  toPlutarch = punsafeCoercePLifted . toPlutarch @inner . coerce @h @(AsHaskell inner)
 
-{- | Helper for writing 'PLiftable' instances. This is /highly/ unsafe, and only
-suitable for internal use!
+  {-# INLINEABLE fromPlutarchRepr #-}
+  fromPlutarchRepr = coerce . fromPlutarchRepr @inner
+
+  {-# INLINEABLE fromPlutarch #-}
+  fromPlutarch p = coerce . fromPlutarch @inner $ punsafeCoercePLifted p
+
+{- | Similar to 'Identity', but at the level of Plutarch. Only needed when
+writing manual instances of 'PLiftable', or if you want to use 'toPlutarch'
+and 'fromPlutarch' directly.
```

Comment:
This is strictly for `coerce`ing Haskell terms. That is running `pcoerceUnsafe :: Term s a -> Term s b` in Haskell level `coerce :: PLifted s a -> PLifted s b`. It'd be nice to note that maybe

Reply:
Actually, I think I can remove this and still make everything work. I'll try this afternoon

Reply:
I've tried but `coerce` yelled at me during `via` deriving